include/runtime/collections/RBTree.h

```diff
@@ -0,0 +1,512 @@
+                lft.right().rootVal(),
+                RBTree(R, lft.right().right(), x, v, rgt.paint(B))));
+      else if (lft.nonEmpty(R) && lft.right().nonEmpty(B) && rgt.isEmpty(BB))
+        return RBTree(
+            B, lft.left(), lft.root(), lft.rootVal(),
+            balance(
+                B, lft.right().left(), lft.right().root(),
+                lft.right().rootVal(),
+                RBTree(R, lft.right().right(), x, v, RBTree())));
+      else
+        return RBTree(c, lft, x, v, rgt);
+    }
+    // Otherwise
+    return RBTree(c, lft, x, v, rgt);
+  }
+
+  // Called only when parent is black or double black
+  static RBTree
+  balance(Color c, RBTree const &lft, T x, V v, RBTree const &rgt) {
+    if (lft.doubledLeft())
+      return RBTree(
+          minusOneColor(c), lft.left().paint(B), lft.root(), lft.rootVal(),
+          RBTree(B, lft.right(), x, v, rgt));
+    else if (lft.doubledRight())
+      return RBTree(
+          minusOneColor(c),
+          RBTree(B, lft.left(), lft.root(), lft.rootVal(), lft.right().left()),
+          lft.right().root(), lft.right().rootVal(),
+          RBTree(B, lft.right().right(), x, v, rgt));
+    else if (rgt.doubledLeft())
+      return RBTree(
+          minusOneColor(c), RBTree(B, lft, x, v, rgt.left().left()),
+          rgt.left().root(), rgt.left().rootVal(),
+          RBTree(
+              B, rgt.left().right(), rgt.root(), rgt.rootVal(), rgt.right()));
+    else if (rgt.doubledRight())
+      return RBTree(
+          minusOneColor(c), RBTree(B, lft, x, v, rgt.left()), rgt.root(),
+          rgt.rootVal(), rgt.right().paint(B));
+    else
+      return RBTree(c, lft, x, v, rgt);
+  }
+
+  bool isEmpty(Color c) const { return isEmpty() && leafColor() == c; }
+
+  bool nonEmpty(Color c) const { return !isEmpty() && rootColor() == c; }
+
+  bool singleton(Color c) const {
+    return !isEmpty() && rootColor() == c && left().isEmpty(B)
+           && right().isEmpty(B);
+  }
+
+  bool onlyLeftChild(Color parentColor, Color childColor) const {
+    return !isEmpty() && rootColor() == parentColor && !left().isEmpty()
+           && left().rootColor() == childColor && right().isEmpty(B);
+  }
+
+  bool onlyRightChild(Color parentColor, Color childColor) const {
+    return !isEmpty() && rootColor() == parentColor && left().isEmpty(B)
+           && !right().isEmpty() && right().rootColor() == childColor;
+  }
+
+  bool doubledLeft() const {
+    return !isEmpty() && rootColor() == R && !left().isEmpty()
+           && left().rootColor() == R;
+  }
+
+  bool doubledRight() const {
+    return !isEmpty() && rootColor() == R && !right().isEmpty()
+           && right().rootColor() == R;
+  }
+
+  RBTree paint(Color c) const {
+    assert(!isEmpty());
+    return RBTree(c, left(), root(), rootVal(), right());
+  }
+
+private:
+  std::shared_ptr<const Node> _root;
+};
+
+template <class T, class V, class F>
+void forEach(RBTree<T, V> const &t, F f) {
```

Comment:
Again nitpicking on best practice here - ideally this would take `f` as a universal reference `F&&`, then use `std::invoke` to call it, and `std::forward` to pass it on to the recursive calls.

Reply:
Addressed.